marklogic-client-api/src/test/java/com/marklogic/client/test/dataservices/BulkOutputCallerNext.java

```diff
@@ -0,0 +1,79 @@
+package com.marklogic.client.test.dataservices;
+
+import com.fasterxml.jackson.databind.node.ObjectNode;
+import com.marklogic.client.dataservices.OutputEndpoint;
+import com.marklogic.client.dataservices.impl.OutputEndpointImpl;
+import com.marklogic.client.document.JSONDocumentManager;
+import com.marklogic.client.io.JacksonHandle;
+import com.marklogic.client.query.DeleteQueryDefinition;
+import com.marklogic.client.query.QueryManager;
+import org.junit.AfterClass;
+import org.junit.BeforeClass;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.ExpectedException;
+
+import java.io.ByteArrayInputStream;
+import java.io.InputStream;
+import java.util.stream.Stream;
+
+import static org.junit.Assert.*;
+
+public class BulkOutputCallerNext {
+    static ObjectNode apiObj;
+    static String apiName = "bulkOutputCallerNext.api";
+    static String scriptPath;
+    static String apiPath;
+    static JSONDocumentManager docMgr;
+
+    private static String collectionName = "bulkOutputCallerNext";
+
+    @Rule
+    public ExpectedException expectedException = ExpectedException.none();
+
+    @BeforeClass
+    public static void setup() throws Exception {
+        docMgr = IOTestUtil.db.newJSONDocumentManager();
+        apiObj = IOTestUtil.readApi(apiName);
+        scriptPath = IOTestUtil.getScriptPath(apiObj);
+        apiPath = IOTestUtil.getApiPath(scriptPath);
+        IOTestUtil.load(apiName, apiObj, scriptPath, apiPath);
+    }
+
+    @Test
+    public void bulkOutputCallerNextTest() throws Exception {
+        String endpointState = "{\"next\":"+1+"}";
+        String workUnit      = "{\"max\":"+5+"}";
+
+        IOTestUtil.writeDocuments(10,collectionName);
+
+        OutputEndpointImpl.BulkOutputCaller bulkCaller = OutputEndpoint.on(IOTestUtil.db, new JacksonHandle(apiObj)).bulkCaller();
+        bulkCaller.setEndpointState(new ByteArrayInputStream(endpointState.getBytes()));
+        bulkCaller.setWorkUnit(new ByteArrayInputStream(workUnit.getBytes()));
+        Stream<InputStream> output = bulkCaller.next();
+        assertNotNull(output);
+        InputStream[] outputArray = output.toArray(size -> new InputStream[size]);
+        assertTrue(outputArray.length == 5);
+        output = bulkCaller.next();
+        assertNotNull(output);
+        InputStream[] outputArray1 = output.toArray(size -> new InputStream[size]);
+        assertTrue(outputArray1.length == 5);
+        assertFalse(outputArray.equals(outputArray1));
+        for(int i=0; i<5; i++) {
+            String data = IOTestUtil.mapper.readValue(outputArray[i], ObjectNode.class).toString();
+            String data1 = IOTestUtil.mapper.readValue(outputArray1[i], ObjectNode.class).toString();
+            assertFalse(data.equals(data1));
+        }
```

Comment:
The test would definitely catch some bad results.

Would the test increase our confidence in the implementation, however, if it tested that every document written was retrieved instead of testing that the document at the same result position are different in the two result sets?

For instance, would the following results test as different

    [doc0, doc1, doc2, doc3, doc4] 
    [doc1, doc2, doc3, doc4, doc0] 

when, in fact, the result sets are the same but in a different order?

Reply:
Yes that's right. Changed the test. It now confirms that each index element is not present in the other list.